src/conduits/RwaInputConduit3.sol

```diff
@@ -0,0 +1,183 @@
+
+    /// @notice GEM token contract address (should ERC20 compliant)
+    DSTokenAbstract public gem;
+    /// @notice PSM contract address for the GEM
+    PsmAbstract public psm;
+
+    /**
+     * @notice `usr` was granted admin access.
+     * @param usr The user address.
+     */
+    event Rely(address indexed usr);
+    /**
+     * @notice `usr` admin access was revoked.
+     * @param usr The user address.
+     */
+    event Deny(address indexed usr);
+    /**
+     * @notice `usr` was granted push access.
+     * @param usr The user address.
+     */
+    event Mate(address indexed usr);
+    /**
+     * @notice `usr` push access was revoked.
+     * @param usr The user address.
+     */
+    event Hate(address indexed usr);
+    /**
+     * @notice `wad` amount of Dai was pushed to `to`
+     * @param to The RwaUrn address
+     * @param wad The amount of Dai
+     */
+    event Push(address indexed to, uint256 wad);
+
+    /**
+     * @notice Define addresses and gives `msg.sender` admin access.
+     * @param _dai Dai token contract address.
+     * @param _to RwaUrn contract address.
+     */
+    constructor(address _dai, address _gem, address _psm, address _to) public {
+        dai = DSTokenAbstract(_dai);
+        gem = DSTokenAbstract(_gem);
+        psm = PsmAbstract(_psm);
+        to = _to;
+
+        require(GemJoinAbstract(psm.gemJoin()).gem() == _gem, "RwaInputConduit3/wrong-gem-for-psm");
+        require(address(psm.dai()) == _dai, "RwaInputConduit3/wrong-dai-for-psm");
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "RwaInputConduit3/not-authorized");
+        _;
+    }
+
+    /**
+     * @notice Grants `usr` admin access to this contract.
+     * @param usr The user address.
+     */
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` admin access from this contract.
+     * @param usr The user address.
+     */
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    /**
+     * @notice Grants `usr` push access to this contract.
+     * @param usr The user address.
+     */
+    function mate(address usr) external auth {
+        may[usr] = 1;
+        emit Mate(usr);
+    }
+
+    /**
+     * @notice Revokes `usr` push access from this contract.
+     * @param usr The user address.
+     */
+    function hate(address usr) external auth {
+        may[usr] = 0;
+        emit Hate(usr);
+    }
+
+    /**
+     * @notice Internal method which:
+     *  - Approve PSM gemJoin Adapter with WAD amount of GEM
+     *  - Swap GEM to DAI through PSM
+     *  - Push DAI to Recipient address (to)
+     * @param wad Amount to swap and push
+     */
+    function _swapAndPush(uint256 wad) internal {
+        gem.approve(address(psm.gemJoin()), wad);
```

Comment:
Instead of approving on-demand, we should probably have an unlimited approval in the constructor to improve gas usage.

Reply:
Done